File: app/src/main/java/com/gee12/mytetroid/ui/node/NodesListAdapter.java
package com.gee12.mytetroid.ui.node;

import android.content.Context;
import android.graphics.Color;
import android.os.Build;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.content.ContextCompat;

import com.gee12.mytetroid.App;
import com.gee12.mytetroid.R;
import com.gee12.mytetroid.model.TetroidNode;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import pl.openrnd.multilevellistview.ItemInfo;
import pl.openrnd.multilevellistview.MultiLevelListAdapter;

public class NodesListAdapter extends MultiLevelListAdapter {

    public interface OnNodeHeaderClickListener {

        /**
         * Обработчик клика на имени ветки.
         *
         * @param node Объект ветки
         */
        void onClick(TetroidNode node, int pos);

        /**
         * Обработчик долгого клика на имени ветки.
         * @param view
         * @param node
         */
        boolean onLongClick(View view, TetroidNode node, int pos);
    }

    private class NodeViewHolder {
        ImageView iconView;
        TextView nameView;
        TextView recordsCountView;
        ImageView arrowView;
        ConstraintLayout headerView;
    }

    private final Context context;
    private final LayoutInflater inflater;
    private OnNodeHeaderClickListener onNodeHeaderClickListener;
    private TetroidNode curNode;

    public NodesListAdapter(Context context, OnNodeHeaderClickListener onNodeHeaderClickListener) {
        super();
        this.context = context;
        this.onNodeHeaderClickListener = onNodeHeaderClickListener;
        this.inflater = LayoutInflater.from(context);
    }

    public void reset() {
        super.setDataItems(new ArrayList<>());
        this.curNode = null;
        notifyDataSetChanged();
    }

    public void setNodeHeaderClickListener(OnNodeHeaderClickListener onNodeHeaderClickListener) {
        this.onNodeHeaderClickListener = onNodeHeaderClickListener;
    }

    @Override
    protected boolean isExpandable(Object object) {
        return ((TetroidNode)object).isExpandable();
    }

    @Override
    protected List<?> getSubObjects(Object object) {
        return ((TetroidNode)object).getSubNodes();
    }

    @Override
    protected Object getParent(Object object) {
        return ((TetroidNode)object).getParentNode();
    }

    @Override
    protected View getViewForObject(Object object, View convertView, ItemInfo itemInfo, int pos) {
        NodeViewHolder viewHolder;
        if (convertView == null) {
            viewHolder = new NodeViewHolder();
            convertView = inflater.inflate(R.layout.list_item_node, null);
            viewHolder.iconView = convertView.findViewById(R.id.node_view_icon);
            viewHolder.nameView = convertView.findViewById(R.id.node_view_name);
            viewHolder.recordsCountView = convertView.findViewById(R.id.node_view_records_count);
            viewHolder.arrowView = convertView.findViewById(R.id.node_view_arrow);
            viewHolder.headerView = convertView.findViewById(R.id.node_view_header);
            convertView.setTag(viewHolder);
        } else {
            viewHolder = (NodeViewHolder) convertView.getTag();
        }
        final View view = convertView;

        final TetroidNode node = (TetroidNode) object;
        // иконка
        if (node.getIcon() != null || node.isCrypted()) {
            viewHolder.iconView.setVisibility(View.VISIBLE);
            viewHolder.iconView.setLayerType(View.LAYER_TYPE_SOFTWARE, null);
            if (!node.isNonCryptedOrDecrypted()) {
                viewHolder.iconView.setImageResource(R.drawable.ic_node_encrypted);
            }
            else if (node.getIcon() != null) {
                viewHolder.iconView.setImageDrawable(node.getIcon());
            } else {
                viewHolder.iconView.setImageResource(R.drawable.ic_node_decrypted);
            }
        } else {
            viewHolder.iconView.setVisibility(View.GONE);
        }
        // название
        String cryptedName = context.getString(R.string.title_crypted_node_name);
        viewHolder.nameView.setText(node.getCryptedName(cryptedName));
        // количество записей в ветке
        if (node.getRecordsCount() > 0 && node.isNonCryptedOrDecrypted()) {
            viewHolder.recordsCountView.setVisibility(View.VISIBLE);
            viewHolder.recordsCountView.setText(String.format(Locale.getDefault(), "[%d]", node.getRecordsCount()));
            viewHolder.nameView.setTextColor(ContextCompat.getColor(context, R.color.colorBaseText));
        } else {
            viewHolder.recordsCountView.setVisibility(View.GONE);
            viewHolder.nameView.setTextColor(ContextCompat.getColor(context, R.color.colorLightText));
        }
        // вьюшка всего заголовка ветки (с иконкой и именем)
//        ((RelativeLayout.LayoutParams)viewHolder.headerView.getLayoutParams()).setMargins(20 * node.getLevel(),0,50,0);
        viewHolder.headerView.setOnClickListener(v -> onNodeHeaderClickListener.onClick(node, pos));
        viewHolder.headerView.setOnLongClickListener(v -> onNodeHeaderClickListener.onLongClick(view, node, pos));
        // стрелка раскрытия/закрытия ветки
        if (itemInfo.isExpandable() && node.isNonCryptedOrDecrypted()) {
            viewHolder.arrowView.setVisibility(View.VISIBLE);
            viewHolder.arrowView.setImageResource(itemInfo.isExpanded() ?
                    R.drawable.ic_arrow_collapse : R.drawable.ic_arrow_expand);
        } else {
            viewHolder.arrowView.setVisibility(View.GONE);
        }
        RelativeLayout.LayoutParams layoutParams = (RelativeLayout.LayoutParams) viewHolder.headerView.getLayoutParams();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            layoutParams.setMarginStart(20 * node.getLevel());
        } else {
            layoutParams.setMargins(20 * node.getLevel(),0, 2,0);
        }
        viewHolder.headerView.setLayoutParams(layoutParams);
//        layoutParams.leftMargin = 20 * node.getLevel();

        // подсветка
        if (curNode == node) {
            convertView.setBackgroundColor(ContextCompat.getColor(context, R.color.colorCurNode));
        } else if (node.isCrypted() && App.IsHighlightCryptedNodes) {
            convertView.setBackgroundColor(App.HighlightAttachColor);
        } else {
            convertView.setBackgroundColor(Color.TRANSPARENT);
        }

        return convertView;
    }

    public void setCurNode(TetroidNode node) {
        this.curNode = node;
    }

    public TetroidNode getCurNode() {
        return curNode;
    }

    public TetroidNode getItem(int position) {
        return (TetroidNode) getListView().getAdapter().getItem(position);
    }

    public void setDataItems(List<?> dataItems) {
        if (dataItems != null) {
            super.setDataItems(dataItems);
        } else {
            super.setDataItems(new ArrayList<>());
        }
    }
}

File: app/src/main/java/com/gee12/mytetroid/ui/node/NodesListAdapter.kt
package com.gee12.mytetroid.ui.node

import android.content.Context
import android.graphics.Color
import android.os.Build
import android.text.Spannable
import android.text.SpannableString
import android.text.style.AbsoluteSizeSpan
import android.text.style.ForegroundColorSpan
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.ContextCompat
import com.gee12.mytetroid.R
import com.gee12.mytetroid.model.TetroidNode
import pl.openrnd.multilevellistview.ItemInfo
import pl.openrnd.multilevellistview.MultiLevelListAdapter
import java.util.*

class NodesListAdapter(
    private val context: Context,
    private val isHighlightCryptedNodes: Boolean,
    private val highlightColor: Int,
    private val onClick: (TetroidNode, Int) -> Unit,
    private val onLongClick: (View, TetroidNode, Int) -> Boolean,
) : MultiLevelListAdapter() {

    private inner class NodeViewHolder {
        lateinit var iconView: ImageView
        lateinit var nameCountView: TextView
        lateinit var arrowView: ImageView
        lateinit var headerView: ConstraintLayout
    }

    private val inflater: LayoutInflater = LayoutInflater.from(context)
    var curNode: TetroidNode? = null

    fun reset() {
        super.setDataItems(ArrayList<Any?>())
        curNode = null
        notifyDataSetChanged()
    }

    override fun isExpandable(obj: Any): Boolean {
        return (obj as TetroidNode).isExpandable
    }

    override fun getSubObjects(obj: Any): List<*> {
        return (obj as TetroidNode).subNodes
    }

    override fun getParent(obj: Any): Any {
        return (obj as TetroidNode).parentNode
    }

    override fun getViewForObject(obj: Any, convertView: View?, itemInfo: ItemInfo, pos: Int): View {
        val view: View
        val viewHolder: NodeViewHolder
        if (convertView == null) {
            viewHolder = NodeViewHolder()
            view = inflater.inflate(R.layout.list_item_node, null)
            viewHolder.iconView = view.findViewById(R.id.node_view_icon)
            viewHolder.nameCountView = view.findViewById(R.id.node_view_name_count)
            viewHolder.arrowView = view.findViewById(R.id.node_view_arrow)
            viewHolder.headerView = view.findViewById(R.id.node_view_header)
            view.tag = viewHolder
        } else {
            view = convertView
            viewHolder = view.tag as NodeViewHolder
        }
        val node = obj as TetroidNode
        // иконка
        if (node.icon != null || node.isCrypted) {
            viewHolder.iconView.visibility = View.VISIBLE
            viewHolder.iconView.setLayerType(View.LAYER_TYPE_SOFTWARE, null)
            if (!node.isNonCryptedOrDecrypted) {
                viewHolder.iconView.setImageResource(R.drawable.ic_node_encrypted)
            } else if (node.icon != null) {
                viewHolder.iconView.setImageDrawable(node.icon)
            } else {
                viewHolder.iconView.setImageResource(R.drawable.ic_node_decrypted)
            }
        } else {
            viewHolder.iconView.visibility = View.GONE
        }
        // название
        val nameWhenEncrypted = context.getString(R.string.title_crypted_node_name)
        val name = node.getCryptedName(nameWhenEncrypted)

        // количество записей в ветке
        if (node.recordsCount > 0 && node.isNonCryptedOrDecrypted) {
            val recordsCount = "[%d]".format(node.recordsCount)
            val nameCount = SpannableString("$name $recordsCount")
            nameCount.setSpan(
                ForegroundColorSpan(ContextCompat.getColor(context, R.color.colorBase2Text)),
                name.length + 1, nameCount.length,
                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            nameCount.setSpan(
                AbsoluteSizeSpan(context.resources.getDimension(R.dimen.font_small).toInt()),
                name.length + 1, nameCount.length,
                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            viewHolder.nameCountView.text = nameCount
            viewHolder.nameCountView.setTextColor(ContextCompat.getColor(context, R.color.colorBaseText))
        } else {
            viewHolder.nameCountView.text = name
            viewHolder.nameCountView.setTextColor(ContextCompat.getColor(context, R.color.colorLightText))
        }
        // вьюшка всего заголовка ветки (с иконкой и именем)
//        ((RelativeLayout.LayoutParams)viewHolder.headerView.getLayoutParams()).setMargins(20 * node.getLevel(),0,50,0);
        viewHolder.headerView.setOnClickListener { onClick(node, pos) }
        viewHolder.headerView.setOnLongClickListener { onLongClick(view, node, pos) }
        // стрелка раскрытия/закрытия ветки
        if (itemInfo.isExpandable && node.isNonCryptedOrDecrypted) {
            viewHolder.arrowView.visibility = View.VISIBLE
            viewHolder.arrowView.setImageResource(if (itemInfo.isExpanded) R.drawable.ic_arrow_collapse else R.drawable.ic_arrow_expand)
        } else {
            viewHolder.arrowView.visibility = View.GONE
        }
        val layoutParams = viewHolder.headerView.layoutParams as RelativeLayout.LayoutParams
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            layoutParams.marginStart = 20 * node.level
        } else {
            layoutParams.setMargins(20 * node.level, 0, 2, 0)
        }
        viewHolder.headerView.layoutParams = layoutParams
//        layoutParams.leftMargin = 20 * node.getLevel();

        // подсветка
        if (curNode == node) {
            view.setBackgroundColor(ContextCompat.getColor(context, R.color.colorCurNode))
        } else if (node.isCrypted && isHighlightCryptedNodes) {
            view.setBackgroundColor(highlightColor)
        } else {
            view.setBackgroundColor(Color.TRANSPARENT)
        }
        return view
    }

    fun getItem(position: Int): TetroidNode {
        return listView.adapter.getItem(position) as TetroidNode
    }

    override fun setDataItems(dataItems: List<*>?) {
        if (dataItems != null) {
            super.setDataItems(dataItems)
        } else {
            super.setDataItems(ArrayList<Any?>())
        }
    }

}